u_fingerprinter_test.go

```diff
@@ -0,0 +1,408 @@
+		for i, originalExtension := range uconn.Extensions {
+			if _, ok := originalExtension.(*UtlsPaddingExtension); ok {
+				// We can't really compare padding extensions in this way
+				continue
+			}
+
+			generatedExtension := generatedUConn.Extensions[i]
+			checkUTLSExtensionsEquality(t, originalExtension, generatedExtension)
+		}
+	}
+
+	fieldsToTest := []string{
+		"Vers", "CipherSuites", "CompressionMethods", "NextProtoNeg", "ServerName", "OcspStapling", "Scts", "SupportedCurves",
+		"SupportedPoints", "TicketSupported", "SupportedSignatureAlgorithms", "SecureRenegotiation", "SecureRenegotiationSupported", "AlpnProtocols",
+		"SupportedSignatureAlgorithmsCert", "SupportedVersions", "KeyShares", "EarlyData", "PskModes", "PskIdentities", "PskBinders",
+	}
+
+	for _, field := range fieldsToTest {
+		compareClientHelloFields(t, field, uconn.HandshakeState.Hello, generatedUConn.HandshakeState.Hello)
+	}
+}
+
+func TestUTLSFingerprintClientHello(t *testing.T) {
+	clientHellosToTest := []ClientHelloID{
+		HelloChrome_58, HelloChrome_70, HelloChrome_83, HelloFirefox_55, HelloFirefox_63, HelloIOS_11_1, HelloIOS_12_1, HelloRandomized, HelloRandomizedALPN, HelloRandomizedNoALPN}
+
+	serverNames := []string{"foobar"}
+
+	for _, clientHello := range clientHellosToTest {
+		for _, serverName := range serverNames {
+			t.Logf("checking fingerprint generated client hello spec against %v and server name: %v", clientHello, serverName)
+			checkUTLSFingerPrintClientHello(t, clientHello, "foobar")
+		}
+	}
+}
+
+func TestUTLSFingerprintClientHelloBluntMimicry(t *testing.T) {
+	serverName := "foobar"
+	var extensionId uint16 = 0xfeed
+	extensionData := []byte("random data")
+
+	specWithGeneric, err := utlsIdToSpec(HelloChrome_Auto)
+	if err != nil {
+		t.Errorf("got error: %v; expected to succeed", err)
+	}
+	specWithGeneric.Extensions = append(specWithGeneric.Extensions, &GenericExtension{extensionId, extensionData})
+
+	uconn := UClient(&net.TCPConn{}, &Config{ServerName: serverName}, HelloCustom)
+
+	if err := uconn.ApplyPreset(&specWithGeneric); err != nil {
+		t.Errorf("got error: %v; expected to succeed", err)
+	}
+	if err := uconn.BuildHandshakeState(); err != nil {
+		t.Errorf("got error: %v; expected to succeed", err)
+	}
+
+	f := &Fingerprinter{}
+	_, err = f.FingerprintClientHello(uconn.HandshakeState.Hello.Raw)
+	if err == nil {
+		t.Errorf("expected error generating spec from client hello with GenericExtension")
+	}
+
+	f = &Fingerprinter{AllowBluntMimicry: true}
+	generatedSpec, err := f.FingerprintClientHello(uconn.HandshakeState.Hello.Raw)
+	if err != nil {
+		t.Errorf("got error: %v; expected to succeed", err)
+	}
+
+	for _, ext := range generatedSpec.Extensions {
+		if genericExtension, ok := (ext).(*GenericExtension); ok {
+			if genericExtension.Id == extensionId && bytes.Equal(genericExtension.Data, extensionData) {
+				return
```

Comment:
checking just the first extension is not enough

Reply:
I'm not sure I follow here - this test is to ensure that the generic extension exists in the generated spec with `AllowBluntMimicry` set. What additional checks would you want here/can you explain the kind of input this wouldn't properly handle?

Reply:
To be (hopefully) more clear: this test manually creates a single generic extension in the client hello which is to be fingerprinted. That generic extension has data = `extensionData`. This for loop is solely interested in finding that the generated spec has had that generic extension with `extensionData` included